test/ar_gateway_middleware_tests.erl

```diff
@@ -0,0 +1,577 @@
+-module(ar_gateway_middleware_tests).
+
+-include_lib("eunit/include/eunit.hrl").
+-include("src/ar.hrl").
+
+-define(MOCK_DOMAIN, <<"gateway.test">>).
+-define(MOCK_CUSTOM_DOMAIN, <<"custom.domain.test">>).
+-define(MOCK_ENV, #{ gateway => {?MOCK_DOMAIN, [?MOCK_CUSTOM_DOMAIN]} }).
+
+-define(MOCK_TXID, <<
+	84, 149, 29, 45, 202, 49, 52, 3, 105, 79, 42, 68, 153, 32, 130, 193,
+	161, 80, 254, 205, 177, 34, 211, 210, 234, 32, 253, 228, 43, 175, 198, 13
+>>).
+
+-define(MOCK_BLOCKID, <<
+	169, 43, 214, 105, 100, 19, 171, 187, 255, 185, 104, 174, 45, 225, 210, 86,
+	75, 211, 187, 148, 156, 59, 69, 31, 77, 165, 56, 190, 20, 75, 14, 207,
+	212, 134, 4, 201, 13, 244, 65, 73, 236, 174, 28, 174, 27, 250, 30, 246
+>>).
+
+-define(MOCK_TXLABEL, <<"ez63a7fvuu3l">>).
+
+-define(MANIFEST_INDEX_TXID, <<
+	92, 107, 206, 33, 17, 203, 162, 58, 142, 193, 82, 199, 205, 201, 190, 14,
+	54, 5, 77, 26, 132, 131, 130, 249, 240, 18, 210, 37, 237, 159, 110, 108
+>>).
+
+-define(MANIFEST_DOG_TXID, <<
+	158, 115, 11, 58, 243, 223, 99, 147, 125, 173, 58, 194, 61, 208, 0, 46,
+	32, 73, 227, 78, 1, 176, 210, 161, 173, 22, 237, 195, 181, 19, 68, 151
+>>).
+
+execute_test_() ->
+	{foreach, fun setup/0, fun teardown/1, [
+		{"redirect root to labeled domain", fun redirect_root_to_labeled_domain/0},
+		{"serve correctly labeled request", fun serve_correctly_labeled_request/0},
+		{"forward other requests to node api", fun forward_other_requests_to_node_api/0},
+		{"render manifest listing", fun render_manifest_listing/0},
+		{"redirect manifest to index", fun redirect_manifest_to_index/0},
+		{"serve manifest subpath", fun serve_manifest_subpath/0},
+		{"handle multi-segment subpaths", fun handle_multi_segment_subpaths/0},
+		{"redirect badly labeled manifest subpaths", fun redirect_badly_labeled_manifest_subpaths/0},
+		{"return 421 on bad manifest", fun return_421_on_bad_manifest/0},
+		{"return 404 on non-existent subpath", fun return_404_on_non_existent_subpath/0},
+		{"return 400 on subpath of non-manifest", fun return_400_on_subpath_of_non_manifest/0},
+		{"serve manifest subpath on custom domain", fun serve_manifest_subpath_on_custom_domain/0},
+		{"forward non-existent subpaths to node api", fun forward_non_existent_subpaths_to_node_api/0}
+	]}.
+
+setup() ->
+	meck:new(ar_storage),
+	meck:new(ar_tx_search),
+	meck:new(ar_domain),
+	meck:expect(
```

Comment:
You can do `meck:new(ar_domain, [passthrough])`, then you do not have to redefine methods you do not want to mock.